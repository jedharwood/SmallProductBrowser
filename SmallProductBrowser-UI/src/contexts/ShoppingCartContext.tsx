import { createContext, useContext, useState } from 'react';
import type { ReactNode } from 'react';
import type { Product } from '../types/product';

interface ShoppingCartItem {
  product: Product;
}

interface ShoppingCartContextType {
  cartItems: ShoppingCartItem[];
  addToCart: (product: Product) => void;
  getTotalItems: () => number;
}

const ShoppingCartContext = createContext<ShoppingCartContextType | undefined>(
  undefined
);

export const useShoppingCart = () => {
  const context = useContext(ShoppingCartContext);
  if (!context) {
    throw new Error(
      'useShoppingCart must be used within a ShoppingCartProvider'
    );
  }
  return context;
};

interface ShoppingCartProviderProps {
  children: ReactNode;
}

export const ShoppingCartProvider = ({
  children,
}: ShoppingCartProviderProps) => {
  const [cartItems, setCartItems] = useState<ShoppingCartItem[]>([]);

  const addToCart = (product: Product) => {
    const updatedCart = [...cartItems, { product }];
    setCartItems(updatedCart);
  };

  const getTotalItems = () => {
    return cartItems.length;
  };

  const value: ShoppingCartContextType = {
    cartItems,
    addToCart,
    getTotalItems,
  };

  return (
    <ShoppingCartContext.Provider value={value}>
      {children}
    </ShoppingCartContext.Provider>
  );
};
