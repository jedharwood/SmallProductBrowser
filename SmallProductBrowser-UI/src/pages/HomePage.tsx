import ProductCard from '../components/ProductCard';
import { useGetProducts } from '../hooks/useProducts';
import type { Product } from '../types/product';

const HomePage = () => {
  const { data: productsData, isLoading, error } = useGetProducts();
  
  if (isLoading) {
    return (
      <div className="min-h-screen bg-gray-100 flex items-center justify-center">
        <div className="text-lg font-semibold">Loading products...</div>
      </div>
    );
  }

  if (error) {
    return (
      <div className="min-h-screen bg-gray-100 flex items-center justify-center">
        <div className="text-red-600 text-lg font-semibold">
          Error loading products: {error.message}
        </div>
      </div>
    );
  }

  if (productsData?.products?.length === 0) {
    return (
      <div className="text-center py-12">
        <p className="text-gray-500 text-lg">No products found.</p>
      </div>
    );
  }

  return (
    <div className="flex flex-col gap-4">
      {productsData?.products?.map((product: Product) => (
        <ProductCard 
            title={product.title}
            price={product.price}
            stock={product.stock}
            thumbnail={product.thumbnail}
            key={product.id}
          />
      ))}
    </div>
  );
};

export default HomePage;
