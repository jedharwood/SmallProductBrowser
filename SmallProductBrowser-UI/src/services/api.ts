import type { Product, ProductsResponse } from '../types/product';

const API_BASE_URL = 'https://localhost:7168';

export const productsApi = {
  getProducts: async (
    search?: string,
    page?: number
  ): Promise<ProductsResponse> => {
    const params = new URLSearchParams();
    if (search) params.append('search', search);
    if (page) params.append('page', page.toString());
    const queryString = params.toString();

    const url = `${API_BASE_URL}/api/products${queryString ? `?${queryString}` : ''}`;
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error('Failed to fetch products');
    }
    return response.json();
  },

  getProduct: async (id: number): Promise<Product> => {
    const response = await fetch(`${API_BASE_URL}/api/products/${id}`);
    if (!response.ok) {
      throw new Error('Failed to fetch product');
    }
    return response.json();
  },
};
