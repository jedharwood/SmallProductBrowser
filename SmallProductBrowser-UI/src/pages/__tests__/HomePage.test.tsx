/// <reference types="@testing-library/jest-dom" />
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ShoppingCartProvider } from '../../contexts/ShoppingCartContext';
import HomePage from '../HomePage';
import { productsApi } from '../../services/api';
import type { ProductsResponse } from '../../types/product';

vi.mock('../../services/api', () => ({
  productsApi: {
    getProducts: vi.fn(),
  },
}));

const TestWrapper = ({ children }: { children: React.ReactNode }) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        staleTime: 0,
        gcTime: 0,
      },
    },
  });

  return (
    <QueryClientProvider client={queryClient}>
      <ShoppingCartProvider>
        <BrowserRouter>{children}</BrowserRouter>
      </ShoppingCartProvider>
    </QueryClientProvider>
  );
};

const mockProductsData: ProductsResponse = {
  products: [
    {
      id: 1,
      title: 'iPhone 15',
      description: 'Latest iPhone',
      price: 999.99,
      stock: 10,
      images: ['iphone.jpg'],
      thumbnail: 'iphone-thumb.jpg',
    },
    {
      id: 2,
      title: 'Samsung Galaxy',
      description: 'Android phone',
      price: 899.99,
      stock: 5,
      images: ['samsung.jpg'],
      thumbnail: 'samsung-thumb.jpg',
    },
  ],
  total: 2,
  skip: 0,
  limit: 10,
};

const mockGetProducts = vi.mocked(productsApi.getProducts);

describe('HomePage', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });
  afterEach(() => {
    vi.resetAllMocks();
  });

  it('should render products when loaded', async () => {
    mockGetProducts.mockResolvedValueOnce(mockProductsData);

    const { container } = render(
      <TestWrapper>
        <HomePage />
      </TestWrapper>
    );

    await waitFor(() => {
      expect(screen.getByText('iPhone 15')).toBeInTheDocument();
      expect(screen.getByText('Samsung Galaxy')).toBeInTheDocument();
    });

    expect(mockGetProducts).toHaveBeenCalledTimes(1);
    expect(mockGetProducts).toHaveBeenCalledWith(undefined, 1);

    expect(container).toMatchSnapshot();
  });
});
