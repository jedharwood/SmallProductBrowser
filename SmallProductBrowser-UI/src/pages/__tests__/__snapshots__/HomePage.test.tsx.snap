// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomePage > should render products when loaded 1`] = `
<div>
  <div
    class="flex flex-col gap-4"
  >
    <div
      class="bg-white rounded-lg shadow-md p-6"
    >
      <div
        class="flex flex-col sm:flex-row gap-4"
      >
        <div
          class="flex-1"
        >
          <input
            aria-label="Search products"
            class="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-blue-500 focus:border-transparent outline-none"
            placeholder="Search products..."
            type="text"
            value=""
          />
        </div>
        <div
          class="flex gap-2"
        >
          <button
            aria-label="Clear search input"
            class="px-6 py-2 bg-gray-500 text-white rounded-lg hover:bg-gray-600 focus:ring-2 focus:ring-gray-500 focus:ring-offset-2 transition-colors"
            disabled=""
          >
            Clear
          </button>
        </div>
      </div>
    </div>
    <a
      data-discover="true"
      href="/product/1"
    >
      <div
        class="bg-white rounded-lg shadow-md overflow-hidden hover:shadow-lg cursor-pointer"
      >
        <div
          class="grid grid-cols-3 gap-4 p-4"
        >
          <div
            class="col-span-2 flex flex-col justify-between"
          >
            <div>
              <h3
                class="text-xl font-semibold text-gray-900 mb-2"
              >
                iPhone 15
              </h3>
            </div>
            <div
              class="space-y-3"
            >
              <div
                class="flex items-center justify-between"
              >
                <span
                  class="text-2xl font-bold text-green-600"
                >
                  £
                  999.99
                </span>
              </div>
              <span
                class="inline-block font-semibold rounded-full px-2 py-1 text-x bg-green-100 text-green-800"
              >
                10 in stock
              </span>
            </div>
          </div>
          <div
            class="col-span-1"
          >
            <img
              alt="An image of iPhone 15"
              class="h-full w-auto object-contain rounded-lg"
              src="iphone-thumb.jpg"
            />
          </div>
        </div>
      </div>
    </a>
    <a
      data-discover="true"
      href="/product/2"
    >
      <div
        class="bg-white rounded-lg shadow-md overflow-hidden hover:shadow-lg cursor-pointer"
      >
        <div
          class="grid grid-cols-3 gap-4 p-4"
        >
          <div
            class="col-span-2 flex flex-col justify-between"
          >
            <div>
              <h3
                class="text-xl font-semibold text-gray-900 mb-2"
              >
                Samsung Galaxy
              </h3>
            </div>
            <div
              class="space-y-3"
            >
              <div
                class="flex items-center justify-between"
              >
                <span
                  class="text-2xl font-bold text-green-600"
                >
                  £
                  899.99
                </span>
              </div>
              <span
                class="inline-block font-semibold rounded-full px-2 py-1 text-x bg-green-100 text-green-800"
              >
                5 in stock
              </span>
            </div>
          </div>
          <div
            class="col-span-1"
          >
            <img
              alt="An image of Samsung Galaxy"
              class="h-full w-auto object-contain rounded-lg"
              src="samsung-thumb.jpg"
            />
          </div>
        </div>
      </div>
    </a>
  </div>
</div>
`;
